[KeyVault] How to fill the form of the Deploy button (second attempt) (#7432)
Azure/azure-sdk-for-js: sdk/keyvault/TEST_RESOURCES_README.md

> # KeyVault Test Resources

Author: You're right! I will do this now 😄

> # KeyVault Test Resources

Author: I don't think we can have one file for all packages though. Some packages have a wide variety of generated resources.

> # KeyVault Test Resources

Reviewer: See my other comment that outlines bullet points for describing the `baseName` ... Parameters to these templates are common and the full list that we use today is: 

* `BaseName` 
* `TestApplicationId` 
* `TestApplicationSecret` 
* `TestApplicationOid`
* `TenantID`

The provisioner app ID and secret are used to authenticate the `New-TestResources.ps1` script, not really required for the ARM templates.

In cases where we aren't using the AAD app in the resource we don't strictly need those parameters in the ARM template. 

As yet [we haven't needed extra parameters in our ARM templates](https://github.com/Azure/azure-sdk-for-js/blob/master/eng/pipelines/templates/jobs/archetype-sdk-integration.yml#L71) beyond those above.

> # KeyVault Test Resources

Author: @danieljurek I understand that from the parameters perspective, but the purchased resources can vary drastically between one ARM template and another. How do we factor that in?

The KeyVault packages provide a common [Azure Resource Manager template][KVARM] to help developers set up the resources needed
to run all of the integration tests in each of the KeyVault clients. This ARM template is used automatically by the CI pipelines
that run during pull requests, just as much as it is used to help developers deploy their own resources on their personal
Azure accounts for manual testing purposes.

If you wish to use this ARM template to deploy resources on your own Azure personal or business account, you can click
the following deploy button:

[![](http://azuredeploy.net/deploybutton.png)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-sdk-for-js%2Fmaster%2Fsdk%2Fkeyvault%2Ftest-resources.json)

> Keep in mind that deploying these resources constitutes a purchase of Azure services that will be billed to your active account.

Reviewer 2: I don't like this wording. It's vaguely legalese and therefore frightening. @richardpark-msft what's your proposal? I can't find it...

> Keep in mind that deploying these resources constitutes a purchase of Azure services that will be billed to your active account.

Reviewer 3: @bterlson  - I didn't unravel the legal-ese, just simplified the sentence:

> Deploying these resources constitutes a purchase of Azure services that will be billed to your active account.

With that said, in this [comment](https://github.com/Azure/azure-sdk-for-js/pull/7432#discussion_r382225823) I mentioned that we don't need to be redundant with the T&C's that pop up when you do the actual deploy in the portal.

I think we could use the same argument here - it's clear from the portal itself that you're purchasing something. So we probably don't need to mention it.

> Keep in mind that deploying these resources constitutes a purchase of Azure services that will be billed to your active account.

Reviewer 2: +1, the portal does seem to make things super obvious.

## What resources will be created

The [ARM template][KVARM] used by the Deploy Button will create:

- A single standard KeyVault, on the provided location, under the provided tenant, with a name based on the provided "Base Name".

You can read more about KeyVault pricing details [here](https://azure.microsoft.com/en-us/pricing/details/key-vault/).

## The form to deploy the KeyVault resources

Once you click the "Deploy to Azure" button, you will be prompted to fill a form.
In this guide we will go through the necessary steps to fill this form for the KeyVault clients for JavaScript.

### Logging in

Clicking the button will make Azure attempt to authenticate your Azure account. Make sure to log in using the Azure account that you want to be billed through.
If you want to **create a new account**, you may follow this guide: [Create an Azure account][CreateAzureAccount].

### The "Subscription" field

The first field you will encounter is the "Subscription" field.
It will have a drop down menu where you will be able to search for the specific Azure subscription
you may want to use as the host of the resources you will create.

> **This field is required.**

Author: The ARM templates currently don't have a lot of information to help users follow through how to fill the properties. Although I agree we should make them better, discussing and following up on feedback on a Markdown file seems like a better initial approach to me.

> **This field is required.**

Reviewer 3: I agree with you in principle - the reason I think we should consider moving more of this into the template is because you'd have to flip back and forth between this readme and the actual portal. It's better to keep them in the one spot.

I just clicked through to get the full experience (love that button!) and TBH it doesn't seem all that confusing for the common properties.

Some things that I realize though:
* You should be able to [derive the tenant ID from their chosen subscription](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-resource#subscription), right? So that's a parameter you could get rid of.
* Each parameter offers this:
  ```json
  "metadata": {
        "description": "The application client ID used to run tests."
   }
  ```
  And that "Description" field shows up in the tooltip directly next to the parameter in the deployment page. Perhaps we should enhance those a bit more so the user can just click the deploy button and go? 
  I know this is bad but I think quite a few people are just going to do that and not even read these documents.

> **This field is required.**

Author: Interestingly enough, I can't add metadata for some fields, like subscription, resource group and location.

> **This field is required.**

Author: I can't add links either.

> **This field is required.**

Author: I have added more information in the properties that I was able to change from the ARM template! Let me know if this is better.

### The "Resource Group" field.

Having picked a subscription, the resource groups available for that subscription
will be pre-filled in a drop-down menu. You can either pick one resource group from this menu,
or click on the "Create new" link right under the drop-down menu to create a new resource group.
You can read more about Resource groups [here][ResourceGroups].

**This field is required.**

> ### The "Location" field.

Author: the location is derived from the resource group in the ARM template, but the portal doesn't seem to understand that: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/test-resources.json#L26

> ### The "Location" field.

Reviewer 3: I'm guessing that's because it's not filled out yet. You could just default it to some name they definitely have to fill in.

So like "fill in a unique name here" or something of that nature.

We can chat about this today if you want. ARM templates are a bit verbose but they have some power behind them.

> ### The "Location" field.

Author: I will make an issue to track this specifically because it needs to be an active effort with Engineering.

> ### The "Location" field.

Author: Issue: https://github.com/Azure/azure-sdk-for-js/issues/7532

> Azure has more global regions than any other cloud provider.

Author: I mean, should we remove it? I like advertising.

> Azure has more global regions than any other cloud provider.

Reviewer 3: @chradek? :)

> Azure has more global regions than any other cloud provider.

Reviewer 4: Might be a good place to link to https://azure.microsoft.com/en-us/global-infrastructure/regions/

**This field is required.**

### The "Base Name" field.

Under "Settings", the "Base Name" will serve us as a prefix
to determine the name of the resources that we will create,
so that you can easily find them later.

This field gets automatically generated for us through our
CI pipelines. It is **required** to replace this default value
with any combination of alphanumeric characters that you might
prefer. Keep in mind that if you have previously generated
resources with the same alphanumeric combination that you're
providing, Azure will likely fail, requiring you to insert
a different value.

### The "Tenant Id" field.

To fill the **Tenant Id** field, you may copy it from the one available on the
[overview page of the Azure Active Directory section of the Azure portal][AADOverview],
or obtain it from the overview page of a recently created Azure Application.
Here's [how to use the portal to create an Azure AD application and service principal that can access resources][HowToAADApp].

**This field is required.**

### The "Test Application Oid" field.

To fill the **Test Application Oid** field, you may get it from the overview page of a recently created Azure application.

> If you don't have an Azure AD application, follow: [How to use the portal to create an Azure AD application and service principal that can access resources][HowToAADApp].

Author: I pressed things on the keyboard 🤣 Nothing concrete! I'll pick a better name.

> If you don't have an Azure AD application, follow: [How to use the portal to create an Azure AD application and service principal that can access resources][HowToAADApp].

Author: Oh, so this is not the ADD Application ID, it's the Object ID of the AAD Application. I know, it's confusing for everyone.

> If you don't have an Azure AD application, follow: [How to use the portal to create an Azure AD application and service principal that can access resources][HowToAADApp].

Reviewer 3: Some more things to think about (from our brainstorming chat earlier):
1. Use the tooltip (which appears to map to the { metadata: description } field from the ARM template.
2. We could also try embedding a link - this would give the user a way to easily find a more extensive set of instructions directly from the tooltip.

> If you don't have an Azure AD application, follow: [How to use the portal to create an Azure AD application and service principal that can access resources][HowToAADApp].

Reviewer: There are 2 ways to access this GUID. 

**From Azure Active Directory -> App Registrations** 

This is where most people would probably land from the linked instructions.

1. Select the app whose Object ID you are looking for. Once you're on the "Overview" page, click the application name in the **Managed application in local directory** 

![image](https://user-images.githubusercontent.com/2158838/75466308-dd11d800-593e-11ea-88f0-885c7c0e8d85.png)

2. The Enterprise Application page for the app will load 
3. Select the Object ID from that page

![image](https://user-images.githubusercontent.com/2158838/75466112-86a49980-593e-11ea-9a35-e601a8c9acac.png)

**From Azure Active Directory -> Enterprise applications** 

1. Select the app whose Object ID you are looking for
2. From the "Overview" page select the Object ID (same as step 3 above)

![image](https://user-images.githubusercontent.com/2158838/75466112-86a49980-593e-11ea-9a35-e601a8c9acac.png)

Once created, go to your application's overview page, and you'll see an "Object ID" property. This is the one that you want to copy and paste on the form.

**This field is required.**

### Terms and conditions

> Make sure to read the Terms and Conditions before purchasing these resources.

Author: It does just pop up in the portal. I'll remove it.

### Verifying your deploy

Once your purchase is completed, a notification badge will appear on the top right of the portal,
indicating that the resource is being created. You can go to your resource group's page,
then go to the **Deployments** section to see a history of all the deployments you've made.

> The deployments that have used an ARM template will say in their names **Microsoft.Template**

Reviewer 3: ```suggestion
Deployments done with an ARM template will have **Microsoft.Template** in their name.
```

If you click in one of them, you'll see at the left "Outputs" as one of the menu options.
If your deployment has succeeded, the outputs will contain the name of the KeyVault that was just created.

## Running the KeyVault client tests

> Once you have the resources ready to run our integration tests,

Reviewer 3: Once you have the resources ready to run our integration tests,
continue reading [how to run our integration tests](./test/README.md).

(maybe)

After your resources have been deployed you can read about [how to run our integration tests](./test/README.md).

Author: I'll take it!
